`box.ctl.on_replication_split_brain_rollback` event

CuriousGeorgiy opened the pull request:
This patch set adds a new `box.ctl.on_replication_split_brain_rollback` event that occurs when an asynchronously committed synchronous transaction gets rolled back by a PROMOTE request to prevent a split brain.

Closes #10460
Closes #11053

CuriousGeorgiy committed:
limbo: add event for asynchronously committed txn rollback

Add a `box.ctl.on_replication_split_brain_rollback` event for notification
of users about the rollback of asynchronously committed synchronous
transactions. This rollback can happen only because of a PROMOTE request
from a different node.

The event is delivered before writing a PROMOTE request for each
asynchronously committed transaction in the limbo that will be rolled back
by the PROMOTE request. This ensures that if the event triggers execution
fails, the PROMOTE request will also fail, ER_SPLIT_BRAIN will be raised,
and the asynchronously committed data will not be lost. The event is not
delivered during recovery of a PROMOTE request.

The event triggers receive a transaction statement iterator similar to
that of other transaction event triggers. The iterator yields all
transaction statement information as it was encoded to the WAL. The
information is decoded from `stmt→row` and pushed onto the Lua stack in a
format similar to the `xlog` module.

The event allows to execute a fully local transaction bypassing the
limbo, which is started before the event triggers execution. The
transaction is committed implicitly after the event triggers execution
finishes if it is not committed explicitly by the user. Failure to
implicitly commit the transaction also raises ER_SPLIT_BRAIN. Trying to run
another transaction other than the provided one (i.e., by explicitly
beginning a new transaction) will always fail.

Closes #10460

@TarantoolBot document
Title: `box.ctl.on_replication_split_brain_rollback` event
Product: Tarantool
Since: 3.4
Root document: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_txn_management/commit/
and https://www.tarantool.io/ru/doc/latest/reference/reference_lua/box_txn_management/atomic/
and https://www.tarantool.io/en/doc/latest/platform/replication/repl_sync/

Added a new `box.ctl.on_replication_split_brain_rollback` event for
notification of users about the rollback of asynchronously committed
synchronous transactions. This rollback can happen only because of a
PROMOTE request from a different node.

The event is delivered before writing a PROMOTE request for each
asynchronously committed transactions in the limbo that will be rolled back
by the PROMOTE request. This ensures that if the event triggers fail, the
PROMOTE request will also fail, ER_SPLIT_BRAIN will be raised, and the
asynchronously committed data will not be lost. The event is not delivered
during recovery of a PROMOTE request.

The event triggers receive a transaction statement iterator similar to
that of other transaction event triggers. The iterator yields all
transaction statement information as it was encoded to the WAL. The
information is decoded from `stmt→row` and pushed onto the Lua stack in a
format similar to the `xlog` module. The iterator returns Lua tables
with the following structure:
```lua
 {
    meta = {
        replica_id = ...,
        lsn = ...,
        term = ...,
    },
    statement = {
        op_type = ...,
        space_id = ...,
        key = ...,
        tuple = ...,
        ops = ...,
        old_tuple = ...,
        new_tuple = ...,
    }
}
```
The values of the `key`, `tuple`, and `ops` fields depend on the operation
type.

The semantics of the event require its triggers to be called multiple times
if the PROMOTE request or the trigger fail. Therefore, the triggers must be
idempotent with respect to transactions. The `lsn` value that is yielded by
the transaction statement iterator is sufficient to uniquely determine
transaction statements.

The event allows to execute a fully local transaction bypassing the
limbo, which is started before the event triggers execution. The
transaction is committed implicitly after the event triggers execution
finishes if it is not committed explicitly by the user. Failure to
implicitly commit the transaction also raises ER_SPLIT_BRAIN. Trying to run
another transaction other than the provided one (i.e., by explicitly
beginning a new transaction) will always fail.

CuriousGeorgiy committed:
limbo: add helper for determining last transaction to be rolled back

In the scope of #10460 we will need to run a trigger for all asynchronously
committed synchronous transactions that are going to be rolled back by a
promote, so let's refactor existing logic for this into a separate helper.

Needed for #10460

NO_CHANGELOG=<refactoring>
NO_DOC=<refactoring>
NO_TEST=<refactoring>

Gerold103 reviewed (commented):
Thanks for the patch! While the coding work is solid, I suggest we step a bit back, and see the thing from afar, trying to address a few design concerns.

CuriousGeorgiy commented on src/box/txn.c:
```diff
@@ -978,6 +978,17 @@ txn_journal_entry_new(struct txn *txn)
 
 		req->approx_len += xrow_approx_len(stmt->row);
 	}
+	if (txn_limbo.is_running_on_split_brain_rollback_triggers) {
+		assert(txn_limbo.is_in_rollback);
+		if (!txn_is_fully_local(txn)) {
+			diag_set(IllegalParams,
+				 "Transactions from the "
+				 "box.ctl.on_replication_split_brain_rollback "
+				 "event trigger must be fully local");
+			return NULL;
+		}
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+	}
```
> Given the uncertainty and depth of the problem I would say this patch needs a design before implementation. Is there a document I could read? Otherwise I suggest we make one, to discuss it.

AFAIC, this patch was already designed and discussed in the scope of the repair queue theme: https://www.notion.so/tarantool/WIP-Automatic-handling-of-async-transactions-from-old-terms-in-CE-and-EE-1882542d61a54e69b41c74f96436ad34?pvs=4#065efdadd8bd4ea6a91c13ef9329230f (didn't have time yet to move it to Confluence).

I agree that the requirement to run an arbitrary number of transactions seems too generic and unnecessary. I would prefer having only one transaction available as you described in `Alternative 2`. I don't think that always starting a transaction will introduce a noticeable overhead — moreover, this code is not the critical path.

However, the behavior will be less transparent and case-specific compared to the current solution.

CuriousGeorgiy replied:
/cc @sergepetrenko @Serpentian

Gerold103 replied:
> I don't think that's really a problem. Why not allow everyone to run fully local transactions during this time?

Because it would mean a bunch of TXNs would slip past the limbo, without being related to the limbo state, but abusing it to ignore the quorum.

Alternatively we could say, that if the limbo is in the rollback state, then it could be treated as "empty", which allows local and async txns to get committed without a quorum indeed. The problem though is what happens if the limbo rollback fails, and some stuff unrelated to the limbo problems was already committed?

That is very edgy. Thanks for the design link btw! Somehow I overlooked in that document, that it would be so intrusive in the code. Specifically the highlighted code chunk that we are discussing. It is a big signal we are doing something wrong, when such an intrusion becomes necessary.

sergepetrenko replied:
> **Alternative 2**
> 
> We start a transaction ourselves unconditionally, force-async. Then user's callback gets executed. The user knows it is already having a txn, and the user's code is supposed to just do whatever operations they need. Then the callback is done, and we try to commit the txn if it is not empty. If the user rolls it back and starts another one inside of the callback, then it won't be force-async, and won't be able to break anything.
> 
> It seems better than the previous solution in a sense that the user doesn't have to leave an open txn hanging. But it is worse in a sense that the user has no choice whether a txn is needed.

I like this variant. It will suit our needs with repair queue just fine and won't introduce any crutches like the current approach. Besides, this trigger is rather low-level, it describes a rarely needed state (similar to `box.ctl.on_schema_init`), so I think this is fine that its semantics differ a bit from other transaction triggers. We could even name this event "internal" to emphasize that it's not intended for general use (although we won't forbid such use, of course).

Besides, we could forbid user-called `box.commit()/box.rollback()` from the trigger similarly to how it was done in de80e0264f7deb58ea86ef85b37b92653a803430

Gerold103 replied:
I would still recommend to reconsider if this should be an event or a trigger. Shame on me I didn't pay enough attention to it in the document 😣. Events semantics is to notify after the fact. Not to change or prevent or piggyback something. The semantics we need here is clearly a trigger - something that can affect the outcome.

For instance, of `box.status` event callback fails, does it prevent the status from changing? Nope. Then why reacting to this `box.ctl.thisnewlongname` event should change the state of anything?

Another example - `space:on_replace` is not an event, it is a trigger, because it can prevent the replace or add more statements to the same txn. Certainly not an event, right?

Does this make sense?

sergepetrenko replied:
It seems there is some misunderstanding here. From what I see it is implemented like a trigger. So is there a problem with naming or something else?

Sure thing `space:on_replace` is a trigger, but it can be set via
```lua
local trigger = require('trigger')
trigger.set('box.space[512].on_replace', 'my_name.on_replace_trigger', function(...) ... end)
```
And here we refer to `box.space[512].on_replace` as an event upon which a trigger is fired. At least I found this terminology  here: https://github.com/tarantool/doc/issues/3988

Gerold103 replied:
Aaaaaah, omg, I was looking at `struct event *on_split_brain_rollback` and was thinking why is it an event like for `box.broadcast()`! Sorry, didn't notice it, then all is ok!

Serpentian replied:
I'm ok with the second variant too

CuriousGeorgiy replied:
Implemented second alternative, as we have discussed.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1520,11 +1698,16 @@ void
 	txn_limbo_create(&txn_limbo);
+	on_split_brain_rollback =
+		event_get("box.ctl.on_replication_split_brain_rollback", true);
+	event_ref(on_split_brain_rollback);
```
I was just following the current naming conventions. This was also discussed in the RFC.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1331,6 +1493,22 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
+	if (recovery_state == FINISHED_RECOVERY &&
+	    txn_limbo_is_owned_by_current_instance(limbo) &&
+	    /*
+	     * A demote request cannot rollback anything, and a rollback request
```
Because it can't rollback anything from the current instance's synchronous queue, just as PROMOTE cannot.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1304,6 +1338,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+	struct xrow_header *row = stmt->row;
+	port_c_create(out);
+	/*
+	 * The iterator returns the following transaction statement information:
+	 *  1. An ordinal request number;
+	 *  2. The term during which the transaction occurred.
+	 *  3. The ID of the transaction;
+	 *  4. The ID of the limbo owner or 0 for local rows that are part of a
+	 *  global transaction;
+	 *  5. The LSN of the request.
+	 *  6. The ID of the space to which the request was applied.
+	 *  7. The request type.
+	 *  8. The argument of the request depending on the operation type
+	 *  9. An optional argument for the UPDATE/UPSERT request, may be nil.
+	 *  10. The tuple which was replaced by the operation, may be nil.
+	 *  11. The tuple created by the operation, may be nil.
```
This design mirrors the design of `on_commit`/`on_rollback` iterators [^1]:
> The iterator has:
> 
> * an ordinal request number
> * the old value of the tuple before the request (nil for an insert request)
> * the new value of the tuple after the request (nil for a delete request)
> * the ID of the space

[^1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_txn_management/on_commit/

CuriousGeorgiy replied:
/cc @sergepetrenko @Serpentian

Gerold103 replied:
Yeah, but in case of on_commit/rollback it is 4 values, while in the new API it is 11 values. This is quite a big difference. We can't expose this kind of multireturn, it is too complex. Even in case of on_commit/rollback I would now say it should have been an object, because now we can't add stuff to it (I know that I did those triggers, but I didn't know any better back then 😭).

sergepetrenko replied:
<img width="946" alt="Снимок экрана 2024-12-17 в 12 34 27" src="https://github.com/user-attachments/assets/16d80b0a-dfc1-45b7-94cc-131c325b3e43" />

Just for the record, here's the format of repair_queue space we've agreed upon. Only `term` is missing, and it's needed, obviously.

From what I can see, `id` and `transaction_id` may be calculated by the trigger itself, no need to pass them,
replica_id, lsn, space_id, operation_type, key/tuple, ops_array, old_tuple, new_tuple may all be passed in a format similar to xlog iterator. Here's the output of `xlog.pairs` when `wal_ext` is enabled:
```
tarantool> for k,v in xlog.pairs('00000000000000000000.xlog') do json = require'json' print(json.encode(k)) print(json.encode(v)) end
1
{"HEADER":{"lsn":1,"replica_id":1,"type":"INSERT","timestamp":1734430078.2496},"BODY":{"space_id":280,"new_tuple":[512,1,"test","memtx",0,{},[]],"tuple":[512,1,"test","memtx",0,{},[]]}}
2
{"HEADER":{"lsn":2,"replica_id":1,"type":"INSERT","timestamp":1734430082.6448},"BODY":{"space_id":288,"new_tuple":[512,0,"pk","tree",{"unique":true},[[0,"unsigned"]]],"tuple":[512,0,"pk","tree",{"unique":true},[[0,"unsigned"]]]}}
3
{"HEADER":{"lsn":3,"replica_id":1,"type":"INSERT","timestamp":1734430089.2591},"BODY":{"space_id":512,"new_tuple":[1,"asdfb"],"tuple":[1,"asdfb"]}}
4
{"HEADER":{"lsn":4,"replica_id":1,"type":"INSERT","timestamp":1734430094.6888},"BODY":{"space_id":512,"new_tuple":[2,"bsdfa"],"tuple":[2,"bsdfa"]}}
5
{"HEADER":{"lsn":5,"replica_id":1,"type":"REPLACE","timestamp":1734430098.0942},"BODY":{"space_id":512,"old_tuple":[2,"bsdfa"],"new_tuple":[2],"tuple":[2]}}
6
{"HEADER":{"lsn":6,"replica_id":1,"type":"UPDATE","timestamp":1734430107.2436},"BODY":{"new_tuple":[1,"asdfb","bababa"],"old_tuple":[1,"asdfb"],"space_id":512,"index_base":1,"key":[1],"tuple":[["=",3,"bababa"]]}}
---
...
```

So I propose the following format:
```yaml
obj = {
    HEADER = {
        lsn = ...,
        replica_id = ...,
        type = ...,
        term = ...
    },
    BODY = {
        space_id = ...,
        key = ...,
        tuple = ...,
        old_tuple = ...,
        new_tuple = ...,
    }
}
```

Gerold103 replied:
I am super super sorry 🙏, but how about we also change the repair queue format to store rows as objects? And only put into the space fields the things like `id`? The indexes anyway can be by json paths, if we need to index here anything. Otherwise I am really scared for extendibility of the format.

Speaking of the iteration: do we really want to make it so binary like `HEADER` and `BODY`? We are not iterating over a file here. And it doesn't reflect the binary structure anyway, because we don't store `term` in xrows, but above it looks like stored in the header.

But I like the rows being objects much more than multireturns 🔥! A couple of suggestions based on the things I wrote above:
- Perhaps flatten the object? So the header and body keys are all in one level? The motivation is what I said about the iterator being over txn statements, not xlogs.
- If not flatten, then perhaps add the term at the same level as `HEADER` and `BODY` keys? Otherwise it isn't what xlog iterator would return.

Thanks for the feedback and the discussion, @sergepetrenko and @CuriousGeorgiy , and apologies once again that I didn't realize those things during the RFC review. I hope not much would need change in the code.

sergepetrenko replied:
> I am super super sorry 🙏, but how about we also change the repair queue format to store rows as objects? And only put into the space fields the things like id? The indexes anyway can be by json paths, if we need to index here anything. Otherwise I am really scared for extendibility of the format.

I agree, this makes sense not to introduce huge space formats with lots of fields. We even have this ticket regarding a similar problem in space `_func`: https://github.com/tarantool/tarantool/issues/9894

I suggest the following format then:
`[id : unsigned, transaction_id : unsigned, row_meta : map, statement : map]`
`row_meta` would contain `replica_id`, `lsn`, `term`.
`statement` would contain the rest: `op_type`, `space_id`, `key`, `tuple`, `old_tuple`, `new_tuple`.

> Perhaps flatten the object? So the header and body keys are all in one level? The motivation is what I said about the iterator being over txn statements, not xlogs.

I'd rather not mix row meta and operation details. It looks strange to pile everything in one level IMO.

> If not flatten, then perhaps add the term at the same level as HEADER and BODY keys? Otherwise it isn't what xlog iterator would return.

Maybe make the  object format similar to space format I propose, then? In this case it would return objects almost in the same format in which they will be stored in the space (`id` and `transaction_id` will be missing, since we agreed the user could calculate them himself). Does it make sense?
```
obj = {
    row_meta = {
        replica_id = ...,
        lsn = ...,
        term = ...,
    },
    statement = {
        op_type = ...,
        space_id = ...,
        key = ...,
        tuple = ...,
        old_tuple = ...,
        new_tuple = ...,
    }
}
```

Gerold103 replied:
Thanks for considering it! Yes, looks way better. I would even suggest to shorten `row_meta` to just `meta`, otherwise top 💪.

Serpentian replied:
Agree with the solution, Sergey proposed

CuriousGeorgiy replied:
> I suggest the following format then:
[id : unsigned, transaction_id : unsigned, row_meta : map, statement : map]

This is out of the scope of this ticket and PR and is more related to the repair queue, but since we started discussing here: AFAIU, transaction identifiers are not stable (i.e., they are affected by snapshots) — I suggest the transaction identifier you are referring  to here should be a monotonically increasing value in this space, rather than `txn->id`.

Implemented @sergepetrenko's variant (`row_meta` -> `meta`) using a MsgPack object.

sergepetrenko replied:
> AFAIU, transaction identifiers are not stable (i.e., they are affected by snapshots) — I suggest the transaction identifier you are referring to here should be a monotonically increasing value in this space, rather than txn->id.

Yes, indeed.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -682,13 +682,10 @@ txn_limbo_write_rollback(struct txn_limbo *limbo, int64_t lsn)
+txn_limbo_last_rollback_entry(struct txn_limbo *limbo, int64_t lsn)
 {
-	assert(limbo->owner_id != REPLICA_ID_NIL || txn_limbo_is_empty(limbo));
-	assert(limbo == &txn_limbo);
```
nit: why don't you want to leave these assertions here?

CuriousGeorgiy replied:
Thanks for noticing, this change was not intended. Returned the assertions.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -1304,6 +1338,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+		return 0;
+	struct txn_limbo_entry *e;
+	rlist_foreach_entry_reverse(e, &limbo->queue, in_queue) {
+		if (!txn_has_flag(e->txn, TXN_WAIT_ACK) ||
```
What about the `TXN_WAIT_SYNC` transactions? It seems we could put them into the repair queue as well. Would this cause some inconsistency?

CuriousGeorgiy replied:
I agree, let's enable this for all asynchronously committed transactions stored in the limbo.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1304,6 +1338,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+	 *  11. The tuple created by the operation, may be nil.
+	 */
+	port_c_add_number(out, entry_it->req_num++);
+	port_c_add_number(out, (uint32_t)txn_limbo.promote_greatest_term);
```
Thanks for noticing, this as a typo — changed the cast to `double`.

CuriousGeorgiy commented on test/replication-luatest/gh_10460_on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,430 @@
+local t = require('luatest')
```
Sorry, we used to name all tests this way in core team — I will start following this style from now. Fixed.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1520,11 +1698,16 @@ void
 	txn_limbo_create(&txn_limbo);
+	on_split_brain_rollback =
+		event_get("box.ctl.on_replication_split_brain_rollback", true);
+	event_ref(on_split_brain_rollback);
```
I see, thanks for pointing this out — fixed.

CuriousGeorgiy committed:
txn: add helper for setting xrow flags on transactions

In the scope of #10460 we will need to set xrow flags on transactions
during recovery rather than simply making them synchronous. Let's refactor
this logic out into a separate helper.

Needed for #10460

NO_CHANGELOG=<refactoring>
NO_DOC=<refactoring>
NO_TEST=<refactoring>

CuriousGeorgiy committed:
box: set TXN_WAIT_SYNC with xrow rather than limbo state during recovery

During recovery we set the `TXN_WAIT_SYNC` flag based on the limbo's state
rather than respecting the xrow's flags. This does not play well with fully
local transactions that were committed using the `TXN_FORCE_ASYNC` flag:
during runtime they bypass the limbo as expected, but during recovery we do
not set this flag, and they are appended to the limbo, if it is not empty,
even though they are not supposed to be, since their xrows does not have
the `wait_sync` flag.

While currently this only affects writes to the `_gc_consumers` space, in
the scope of #10460 we will also need to enable fully local transactions
from the `box.ctl.on_replication_split_brain_rollback` trigger to be
committed bypassing the limbo.

To fix this, during recovery let's set all the required flags related to
synchronous replication while processing the xrow, and ignore the limbo
emptiness, if we are in recovery and processing local transactions (i.e.,
not applying remote transactions).

Closes #11053
Needed for #10460

NO_DOC=<bugfix>

CuriousGeorgiy commented on changelogs/unreleased/gh-10460-on_replication_split_brain_rollback-event.md:
```diff
@@ -0,0 +1,5 @@
+## feature/box
+
+* Added a new `box.ctl.on_replication_split_brain_rollback` event that occurs
```
I was referring to how individual entries are pushed onto the stack:
https://github.com/tarantool/tarantool/blob/ec65c54af009086b6e2fd603cc2095ce3587b220/src/box/lua/xlog.c#L123-L140

CuriousGeorgiy commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,430 @@
+        box.schema.space.create('l', {is_local = true}):create_index('p')
+        box.ctl.promote()
+    end)
+    cg.master:wait_for_downstream_to(cg.replica)
```
Sorry, I forgot to remove your assignment, and you were looking at a WIP version. Should be fine now — does not flake on my machine.

CuriousGeorgiy commented:
Checkpatch output seems to FP for my case:
```console
ERROR:MACRO_ARG_PRECEDENCE: Macro argument 'field_name' may be better as '(field_name)' to avoid precedence issues
#217: FILE: src/box/txn_limbo.c:1383:
+#define encode_request_field(field_name)				    \
+do {									    \
+	if (request.field_name != NULL) {				    \
+		mpstream_encode_str(&stream, #field_name);		    \
+		size_t len = request.field_name##_end - request.field_name; \
+		char *buf = mpstream_reserve(&stream, len);		    \
+		memcpy(buf, request.field_name, len);			    \
+		mpstream_advance(&stream, len);				    \
+	}								    \
+}									    \
+while (0)
```

I couldn't figure out how to make it happy.

Gerold103 commented on src/box/txn.c:
```diff
@@ -1416,6 +1416,12 @@ box_txn_set_timeout(double timeout)
+}
+
+void
+box_txn_set_xrow_flags(uint8_t xrow_flags)
```
The function `box_txn_make_sync()` has `box_` prefix because it is public, available in module.h. For internal txn methods lets better use just `txn_`, without `box_`. It would then be `txn_set_xrow_flags()`. Does that make sense?

CuriousGeorgiy replied:
Yes, I followed this logic, however `txn_` functions accept a `txn` parameter (since they are like a method), while I want this function to handle the case when there is no `txn` transparently. Would you prefer to have a `txn_` function to which we pass `NULL` in some cases? I think it would look weird.

Gerold103 replied:
I think we can safely call it `txn_` and not require an explicit txn parameter. Subjectively to me it feels like in this case the method prefixes are mostly about their visibility and applicability scope. We do already have `txn_detach()`, `txn_begin_ro_stmt()` (it uses `in_txn()` internally), `txn_savepoint_release()`. As much as I would like us to pass txn always explicitly, when it comes to fiber-local data, we seem to be ok.

CuriousGeorgiy replied:
> We do already have `txn_detach()`, `txn_begin_ro_stmt()` (it uses `in_txn()` internally), `txn_savepoint_release()`.

Forgot about those — then I completely agree, fixed.

Gerold103 commented on src/box/txn.c:
```diff
@@ -995,7 +995,8 @@ txn_journal_entry_new(struct txn *txn)
 			txn_set_flags(txn, TXN_WAIT_SYNC | TXN_WAIT_ACK);
-		} else if (!txn_limbo_is_empty(&txn_limbo)) {
+		} else if (recovery_state == FINISHED_RECOVERY &&
+			   !txn_limbo_is_empty(&txn_limbo)) {
```
> During recovery and replication, the `TXN_WAIT_SYNC` must not be set for such transactions.

Why isn't `TXN_FORCE_ASYNC` enough? I don't get this part. No matter recovery or not, if a txn didn't have these flags during WAL write, then it won't get them. Otherwise if it had them, then they should stay. There doesn't seem to be any need to clear this flag on recovery, if it was present before recovery.

CuriousGeorgiy replied:
> No matter recovery or not, if a txn didn't have these flags during WAL write, then it won't get them.

Since we set them during recovery based on the limbo state, it will actually get them. See https://github.com/tarantool/tarantool/issues/11053 (WIP changing this commit from refactoring to bugfix).

Gerold103 replied:
Omg 😱, thanks for noticing and reporting this! Is there a chance we could make this in limbo.c file somehow? Since this behaviour is caused by the limbo, it seems natural if we could also work it around inside the limbo, hm? For example, when add_entry happens or somewhere there 🤔?

CuriousGeorgiy replied:
> Is there a chance we could make this in limbo.c file somehow?

I don't think so: how are we going to distinguish between a transaction for which the xrow has the `wait_sync` flag versus a transaction for which we set the `TXN_WAIT_SYNC` because the limbo was not empty? The problem is essentially that we set the `TXN_WAIT_SYNC` flag too aggressively.

> Since this behaviour is caused by the limbo, it seems natural if we could also work it around inside the limbo, hm?

I don't think this a problem, TBH: we have the logic for setting the synchronous replication flags in `txn_journal_entry_new`, so it's reasonable to keep it all there.

Gerold103 replied:
Thanks for explaining, I see the problem now. You managed to elaborate on that very well, and I think I might have a new idea how we can make it in another way.

Recovery is being processed in box.cc in `wal_stream_apply_dml_row()`. Can we just check those things there and set the force_async flag just there? Before calling `txn_commit_submit()` the function could check if the txn must be force-async.

I want to explain a bit why I am afraid of expanding `txn_journal_entry_new()` - to be completely honest with you, the setting of flags in this function was a sacrifice made for perf, which I am not even sure is worth it anymore. I needed a way how to check whether a txn is sync. For that one has to walk all the statements. And txn-commit function has exactly one place where it happens - `txn_journal_entry_new()` 😣. So I piggybacked the flag setting into this iteration over the statements. Ideally this function should have never changed the txn. Its name is literally "journal entry new", it must only create the journal entry. This was a bad idea I think, from my side. Perhaps I should have better walked the txn statement list twice or at least renamed the function.

However what we could say until now is that at least this function only looks at the txn itself and the limbo. Only 1 global state is involved. The limbo.

With the new commit this function operates on 2 global states: limbo and recovery state. Moreover, it checks the recovery state even after the recovery is surely done. This just looks somehow not right.

And I am trying to find a way to keep the recovery out of it. Seems like all the recovery txns are coming from `wal_stream_apply_dml_row()`. For the xlogs. For the snapshot I think we have something similar, but don't remember where. Perhaps in the snapshot case all is fine anyway.

So, would making this same check in `wal_stream_apply_dml_row()` be enough? If necessary, we could walk the txn statements in that function.

CuriousGeorgiy replied:
Fixed, removed the check from `txn_journal_entry_new`, now we set `TXN_FORCE_ASYNC` during recovery of all transactions that did not go through the limbo at commit time.

Gerold103 replied:
Thank you 🙏.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -682,13 +682,10 @@ txn_limbo_write_rollback(struct txn_limbo *limbo, int64_t lsn)
+txn_limbo_last_rollback_entry(struct txn_limbo *limbo, int64_t lsn)
 {
-	assert(limbo->owner_id != REPLICA_ID_NIL || txn_limbo_is_empty(limbo));
-	assert(limbo == &txn_limbo);
```
In response to Sergey you said you returned the assertions. But you moved them to `txn_limbo_read_rollback()`. Why? Is there a case where they aren't true for `txn_limbo_last_rollback_entry()`?

CuriousGeorgiy replied:
I just wanted to make the least amount of changes for clarity, but I see how it raises more questions — AFAIC, the assertions always hold, moved them to `txn_limbo_last_rollback_entry`.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -38,9 +38,36 @@
+#include "port.h"
+#include "tuple.h"
+#include "mpstream/mpstream.h"
+#include "lua/tuple.h"
```
Hm. Making limbo depend on Lua is a bold move 😁. And we absolutely can't do that, unfortunately or not. Lets take care of the dependencies, so they don't get bloated and entangled too much. The future-us will say "thanks" to the past-us.

I see that you use it for `tuple_to_mpstream()` function. But it doesn't seem to depend on Lua at all. I suggest to follow the Boy-Scout Rule ("leave code better than it was before you"), and move this function to box/tuple.h(.c).

CuriousGeorgiy replied:
Sorry, I didn't pay enough attention to this move — thanks for highlighting, I completely agree. Fixed.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -1306,6 +1337,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+	mpstream_flush(&stream);
+	size_t obj_size = region_used(region) - region_svp;
+	char *obj = xregion_join(region, obj_size);
+	port_c_add_mp_object(out, obj, obj + obj_size, NULL);
```
Why using an mp object instead of just `port_c_add_mp()`? People won't be able to insert this directly to any space (it is a map, not an array). Which means they will have to decode it anyway, to get the individual fields, check if it wasn't inserted already (the callback can be called more than once for the same txns if PROMOTE fails).

Is there a specific reason we use mpobject here?

CuriousGeorgiy replied:
MsgPack objects are more user-friendly than raw MsgPack strings with indexing APIs, and they are already extensively used in our user APIs:
1. Persistent function `takes_raw_args` option to pass arguments as MsgPack objects [^1]
2. net.box `return_raw` option to return results as MsgPack objects [^2]
3. `box.space.test.on_recovery_replace` and `box.space.test.before_recovery_replace` trigger arguments [^3]
4. `box.iproto.decode_packet` return values [^4]

Moreover, MsgPack objects support lazy decoding, so one can skip decoding transactions that are not interesting, as in the case when the trigger is called more than once for the same transactions when PROMOTE fails.

[^1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/func_create/#lua-data.function_options.takes_raw_args
[^2]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/#lua-function.conn.request
[^3]: https://github.com/tarantool/doc/issues/3988
[^4]: https://github.com/tarantool/doc/issues/3827

CuriousGeorgiy replied:
> So when we return port_c_add_mp() to Lua code, it stays as binary strings? I had impression that they get decoded into Lua tables 🫣.

Sorry, I was wrong on that part, they are indeed decoded into Lua tables:
https://github.com/tarantool/tarantool/blob/43aa0bf45ef18f76b312fabf3d3842d81a970bae/src/box/lua/misc.cc#L319-L329

AFAIC, one of the benefits for why MsgPack objects are used in the APIs I mentioned above is that one can insert them into a space if they store an array [^1]:
```lua
box.space.bands:insert(msgpack.object({1, 'The Beatles', 1960}))
```

But in our case we store a map, so we cannot benefit from this. I guess the argument about lazy decoding is not so relevant, as this code is not on the critical path of execution, so I agree — let's just use `port_c_add_mp()`, as you suggested.

[^1]: https://www.tarantool.io/ru/doc/latest/reference/reference_lua/msgpack/#lua-object.msgpack.msgpack_object

Gerold103 replied:
Thanks for the investiation!

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -1306,6 +1337,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+	encode_request_field(key);
+	encode_request_field(tuple);
+	encode_request_field(ops);
+#undef encode_request_part
```
`encode_request_part` is not defined anywhere. Did you mean `encode_request_field`?

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -1306,6 +1337,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+do {									    \
+	if (request.field_name != NULL) {				    \
+		mpstream_encode_str(&stream, #field_name);		    \
+		size_t len = request.field_name##_end - request.field_name; \
+		char *buf = mpstream_reserve(&stream, len);		    \
+		memcpy(buf, request.field_name, len);			    \
+		mpstream_advance(&stream, len);				    \
```
Wouldn't `mpstream_memcpy()` work shorter?

CuriousGeorgiy replied:
You, are right — judging by this code, I had a false impression that `mpstream_memcpy` does not reserve space. Fixed.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -1333,6 +1492,32 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
+		 * for committing the transaction.
+		 */
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (on_split_brain_rollback_run_triggers(limbo,
```
What happens if users would yield in their trigger? Also `box_txn_commit()` yields. AFAIR, `txn_limbo_req_prepare()` was never ready to yields. For example, some actions it does in the end seem important to be done fast. I suggest to move this code block to the end of the function. Or even to `txn_limbo_req_commit()`.

CuriousGeorgiy replied:
> AFAIR, txn_limbo_req_prepare() was never ready to yields.

What do you mean by `ready for yields`? This patchset includes tests that run transactions which make WAL write, hence, yielding — the tests work fine. 

> For example, some actions it does in the end seem important to be done fast.

Looking at this code, I don't see why is has to be done `fast`:
https://github.com/tarantool/tarantool/blob/b24328cd57b7cdf9e425394d1aa86e2be450f658/src/box/txn_limbo.c#L1333-L1341

> I suggest to move this code block to the end of the function.

I don't think this would be good: in case of failure this would require rolling back the actions done in the code block above.

> Or even to `txn_limbo_req_commit()`.

This definitely won't work. We cannot fail in `txn_limbo_req_commit`.

Gerold103 replied:
> What do you mean by ready for yields?

I mean we never intended for this function to yield. Who knows what happens if some other fiber can observe `is_in_rollback` being set, nothing else being changed, and `PROMOTE` in progress?

Anyway, nevermind, it is just me being paranoid again. And you are completely right about everything else, like that it is covered with tests; it has to be done before anything else because it can fail; and that we would have to revert things if move this to the end of `txn_limbo_req_prepare()`.

Thank you for checking!

Gerold103 reviewed (commented):
Thanks for the fixes, @CuriousGeorgiy. It is a huge work, and I am thankful you've addressed my comments so diligently and patiently 🙏. Especially about the iteration format and when to start the txns.

CuriousGeorgiy committed:
box: lift `tuple_to_mpstream` from the Lua library to the box library

The `tuple_to_mpstream` helper function does not depend on Lua, so let's
lift it from the Lua library to the box library.

Needed for #10460

NO_CHANGELOG=<refactoring>
NO_DOC=<refactoring>
NO_TEST=<refactoring>

Gerold103 commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+-- Test that the event iterator returns correct information for different
+-- transaction statements.
+g.test_different_stmts = function(cg)
+    local stmts = {
+        {'replace', {0}, nil, nil, {0}},
+        {'replace', {777, 1}, nil, {777, 0}, {777, 1}},
+        {'insert', {0}, nil, nil, {0}},
+        {'delete', {0}, nil, nil, nil},
+        {'delete', {777}, nil, {777, 0}, nil},
+        {'update', {777}, {{'=', 2, 1}}, {777, 0}, {777, 1}},
+        {'update', {777}, {{'=', 2, 1}, {'=', 3, 1}}, {777, 0}, {777, 1, 1}},
+        {'upsert', {777}, {{'=', 2, 1}}, {777, 0}, {777, 1}},
+        {'upsert', {0}, {{'=', 2, 1}}, nil, {0}},
+    }
```
Have you thought about maybe having them as key-value tables? Like
```
local stmts = {
    {type = 'replace', key = ..., tuple = ... },
    ...
}
```
Tbh, it is quite hard to read at the moment, and you have to pass all those items individually in the code above, while could pass each statement as a key-value table. I feel like this would simplify the code above quite a bit.

CuriousGeorgiy replied:
Thanks, applied your suggestion, it indeed looks better now.

I guess the lesson of this patch series for me is to stop passing so much stuff around 😁

Gerold103 commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+g.test_async_committed_async_tx_in_limbo = function(cg)
+    partition_server(cg.replica)
+    cg.master:exec(function()
+        require('fiber').create(function()
+            box.space.s:replace{0}
+        end)
```
You must be able to use `submit` with synchronous txns too now. I think it was merged to master already.

CuriousGeorgiy replied:
I need to synchronously commit this transaction.

Gerold103 commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+g.test_sync_committed_sync_tx = function(cg)
+    partition_server(cg.replica)
+    cg.master:exec(function()
+        require('fiber').create(function()
+            box.space.s:replace{0}
+        end)
```
Ditto.

CuriousGeorgiy replied:
Ditto.

Gerold103 commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+        local event = 'box.ctl.on_replication_split_brain_rollback'
+        _G.trigger.set(event, 't', function()
+            local msg = "Can't modify data on a read-only instance"
+            _G.trigger_ok= pcall(function()
```
Whitespace missing after `trigger_ok`.

CuriousGeorgiy replied:
Fixed.

Gerold103 commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+        end)
+        t.assert_equals(box.info.synchro.queue.len, 1)
+    end, {cg.replica:get_instance_id()})
+end
```
I have to say - the test coverage is very impressive, I wouldn't even think about some of those cases myself.

Gerold103 reviewed (commented):
Thank you for the fixes! Now I had a closer look at the tests, given that the main patch is basically almost perfect.

CuriousGeorgiy commented on test/replication-luatest/gh_11053_gc_consumer_record_lost_after_promote_and_restart_test.lua:
```diff
@@ -0,0 +1,80 @@
+    cg.replica_set:wait_for_fullmesh()
+    cg.master:exec(function()
+        box.schema.space.create('s', {is_sync = true}):create_index('p')
+        box.ctl.promote()
```
Please note that the elections are disabled by default and that I do not enable them for tests — a promote request does not need quorum.

Gerold103 replied:
Omg, I forgot how horrible the promotion without elections is 😂.

Gerold103 replied:
But then why do you even specify the quorum at all, if its value doesn't matter? It does look confusing in the test.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1306,6 +1336,132 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+	struct txn_limbo_entry *e;
+	struct region *region = &fiber()->gc;
+	rlist_foreach_entry_reverse(e, &limbo->queue, in_queue) {
+		if (e->txn->fiber != NULL)
+			continue;
```
Because we only care about asynchronously committed transactions (which don't have a fiber attached to them), for which the user cannot get the result of the commit. For synchronously committed transactions (i.e., `wait_mode = 'complete'`) the user will get the result of the commit directly.

Gerold103 replied:
But wait, the user can get async commits results too. With `on_commit` trigger. And in `on_rollback` trigger (if the user would set them), the user has no way to see if it was due to split brain. Shouldn't it be all unified regardless of `wait_mode`?

Gerold103 replied:
The repair queue AFAIU was mainly for async transactions, which were committed without a quorum. Not for non-blocking synchro txns. Those anyway can't be confirmed and then rolled back. This type of split-brain handling only works for the "fake split brain" which is caused by async txns.

If we get a split-brain on actual synchronous txns, it only means the quorum wasn't right, and this can't be repaired. Because such txns most likely aren't even in the limbo already.

sergepetrenko replied:
I agree, let's discuss this verbally. And also summon @Serpentian.

CuriousGeorgiy commented on test/replication-luatest/on_replication_split_brain_rollback_event_test.lua:
```diff
@@ -0,0 +1,496 @@
+                    t.assert_equals(n_stmts, counter - 1)
+                end
+            end)
+        box.cfg{replication_synchro_quorum = ''}
+        box.space.s:replace{777, 0}
+        box.cfg{replication_synchro_quorum = 3}
```
Because we need a tuple to exist in the database to test `update` and `upsert` statements.

CuriousGeorgiy commented on src/box/txn_limbo.c:
```diff
@@ -1333,6 +1489,32 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (on_split_brain_rollback_run_triggers(limbo,
+							 req->lsn + 1) != 0 ||
+		    box_txn_commit() != 0) {
```
> What if there are multiple independent triggers? AFAIU, the event supports that. Is it intentional that they will all run in the same txn? 

Yes, it is intentional.

> What if there were a lot of async rows committed - the txn might get quite large.

So what's the problem that you see? От перестановки мест слагаемых сумма не меняется.

> Do we maybe want to run it per each invocation of the callback?

What if one of the triggers fails, while we commit the transactions for other triggers — how we are going to handle that? I think the current semantics is the most predictable and well-defined.

> And whatever is decided, lets mention it in the doc.

Reworded the commit message to explicitly use event trigger_s_ execution.

sergepetrenko replied:
I think we should ban issuing`box.begin()/box.commit()` from the trigger callback.
Or at least mention in the doc that these explicit box.begin()/box.commit() will break the user's expectations about the trigger atomicity.

Other than that - I don't see a problem in having a single large txn for everything done in the trigger.

Gerold103 commented:
@sergepetrenko and @Serpentian , please, join the review. I have no major comments left. We are just discussing last nits with Georgy.

sergepetrenko reviewed (commented):
Thanks for the patch!

What bothers me is
1. the trigger name
2. that the trigger is ran only on PROMOTE, but not on usual ROLLBACK (although there the user also doesn't know the state of a wait='submit' transaction).
3. that some transactions (committed synchronously) are not pushed to the trigger

Let's discuss this on a call.

Gerold103 commented on changelogs/unreleased/gh-10460-on_replication_split_brain_rollback-event.md:
```diff
@@ -0,0 +1,6 @@
+## feature/box
+
+* Added a new `box.ctl.on_replication_split_brain_rollback` event that occurs
```
Good discussion, and I am very sorry that we have to bring this all up at such a late stage 😭, Georgy did a wonderful job in the code, I hope we won't have to change much.

Speaking of names - this is indeed not really a split-brain, when we rollback not committed synchronous transactions. So I like more with the `promote` suffix. Or even deeper - `on_leader_change_tail_rollback`.
